vpr/src/place/place_util.h

```diff
@@ -205,5 +213,42 @@ int get_initial_move_lim(const t_placer_opts& placer_opts, const t_annealing_sch
 ///@brief Returns the standard deviation of data set x.
 double get_std_dev(int n, double sum_x_squared, double av_x);
 
+//Placement checkpoint
+/**
+ * @brief Data structure that stores the placement state and saves it as a checkpoint.
+ *
+ * The placement checkpoints are very useful to solve the problem of critical 
+ * delay oscillations, expecially very late in the annealer. 
+ *
+ *   @param cost The weighted average of the wiring cost and the timing cost.
+ *   @param block_locs saves the location of each block
+ *   @param cpd Saves the critical path delay of the current checkpoint
+ *   @param valid a flag to show whether the current checkpoint is initialized or not
+ */
+class t_placement_checkpoint {
+  private:
+    vtr::vector_map<ClusterBlockId, t_block_loc> block_locs;
+    float cpd;
+    float sTNS;
+    float sWNS;
+    bool valid = false;
+    t_placer_costs costs;
+
+  public:
+    void save_placement(const t_placer_costs& costs, const float& cpd);
```

Comment:
A brief comment on what it saves would be useful:  gets the block locations from the placement context?  Passed in cpd and costs (not recomputed or checked to be correct).
Same for restore placement: restores to placement_context?

Reply:
Added a comment to explain what each function is doing. We pass cpd and costs and save it in the checkpoint to be able to compare with it  after the quench and see whether to restore or not.